pkg/process/events/store.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2016-present Datadog, Inc.
+
+package events
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-go/v5/statsd"
+	"go.uber.org/atomic"
+
+	ddconfig "github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/process/events/model"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// Ensure in compile time that the RingStore satisfies the Store interface
+var _ Store = &RingStore{}
+
+// Store defines an interface to store process events in-memory
+type Store interface {
+	// Run starts the store
+	Run()
+	// Stop stops the store
+	Stop()
+	// Push sends an event to be stored. An optional channel can be passed to acknowledge when the event is successfully written
+	Push(*model.ProcessEvent, chan bool)
+	// Pull fetches all events in the store that haven't been consumed yet
+	Pull(context.Context, time.Duration) ([]*model.ProcessEvent, error)
+}
+
+// ringNode represents an event stored in the RingStore buffer
+// It can eventually hold more data as lastUpdate, event hash etc
+type ringNode struct {
+	data *model.ProcessEvent
+}
+
+// pushRequest is a request sent to the internal RingBuffer routine to add an event to its buffer
+type pushRequest struct {
+	event *model.ProcessEvent
+	done  chan bool
+}
+
+// pullRequest is a request sent to the internal RingBuffer routine to fetch events from its buffer
+type pullRequest struct {
+	results chan []*model.ProcessEvent
+}
+
+// RingStore implements the Store interface using a ring buffer
+// The buffer is accessed by a single routine so it doesn't need to be protected by a mutex
+// It holds two pointers, head and tail, to access the ring buffer
+// head points to the oldest event in the buffer, where data should be consumed from
+// tail points to the node where the next event will be inserted into
+// head = tail if
+//		* the store is empty, in which case the underlying ringNode doesn't have any data
+// 		* the store is full. Subsequent Push operations override the data pointed by head and move both head and tail
+//		to the next position
+type RingStore struct {
+	head     int
+	tail     int
+	buffer   []ringNode
+	maxItems int
+
+	dropHandler EventHandler // applied to an event before it's dropped. Used for test's purposes
+	pushReq     chan *pushRequest
+	pullReq     chan *pullRequest
+
+	wg   sync.WaitGroup
+	exit chan struct{}
+
+	statsdClient  statsd.ClientInterface
+	statsInterval time.Duration
+	expiredInput  *atomic.Int64 // how many events have been dropped due to a full pushReq channel
+	expiredBuffer *atomic.Int64 // how many events have been dropped due to a full buffer
+}
+
+// readPositiveInt reads a config stored in the given key and asserts that it's a positive value
+func readPositiveInt(key string) (int, error) {
+	i := ddconfig.Datadog.GetInt(key)
+	if i <= 0 {
+		return 0, errors.New("invalid setting. " + key + " should be > 0")
```

Comment:
nit: consider using `fmt.ErrorF(..)` to do the string formatting for the err instead.

Reply:
Also, consider stronger wording in the error msg here. `should` -> `must`

Reply:
Updated in https://github.com/DataDog/datadog-agent/pull/12159/commits/5f95d06ed2feb6fecd77eb0f7c7086893b3f5b50.